Conditional local statement

e-kayrakli opened the pull request:
Modifies the grammar for `local` statement to allow conditional `local` blocks.

With this PR, the `local` statement is identical to the `serial` statement grammar-wise. See issue #6392 for more detailed discussion.

Also fixes some of the tests with old grammar I was able to find with naive grepping. I suspect that there are more that I missed.

TODO:
- [x] full paratest (something I cannot do myself)

e-kayrakli commented:
@bradcray  and @benharsh -- here is what I have for this.

It seemed to be working with much less effort than I was anticipating, which makes me a bit uncomfortable. I hope I am not missing anything.

benharsh commented:
Can you add a test case for a program that would fail to run if not for the conditional feature?

e-kayrakli commented:
@benharsh -- as in a test case that uses the old grammar?

Like `local foo()`(old) instead of `local do foo()`(new)

benharsh commented:
I meant something that uses the conditional to avoid a local block around remote code. Something along these lines:

```chpl
proc inc(x) {
  local x.locale == here do x += 1;
}
var A = 1;
inc(A);

// should enter non-local branch
on Locales[numLocales-1] do inc(A);
```

Basically a motivating use-case for this feature.

e-kayrakli commented:
Ah, I see. Will do!

bradcray commented on compiler/parser/chapel.ypp:
```diff
@@ -480,7 +480,8 @@ stmt:
 | TLABEL ident stmt                    { $$ = buildLabelStmt($2, $3); }
-| TLOCAL stmt                          { $$ = buildLocalStmt($2); }
+| TLOCAL expr do_stmt                  { $$ = buildLocalStmt($2, $3); }
+| TLOCAL do_stmt                       { $$ = buildLocalStmt(new SymExpr(gTrue), $2); }
```
Continuing our conversation in the top-level notes, I think I'd expect this to simply be `buildLocalStmt($2)` in order to avoid the early-compiler code bloat that you noted.

e-kayrakli commented:
I think this is ready for another review/paratest.

P.S. Commit history is a bit messy with some back-and-forths, maybe this should be squash-merged. OTOH there are parser rebuilds, which may not be a good idea to squash with other commits.

e-kayrakli commented:
@benharsh -- you can send me an email if the list of syntax errors in tests is too long :)

benharsh commented:
I'll track testing here:

- [ ] full local
- [ ] full no-local
- [ ] release/examples gasnet

benharsh commented:
Can you also update the documentation in doc/rst/technotes/local.rst ?

benharsh commented:
I think you missed some local statements in the following files:
- modules/dists/StencilDist.chpl
- test/optimizations/localon/sublocale.chpl
- test/release/examples/benchmarks/ssca2/SSCA2_Modules/SSCA2_RMAT_graph_generator.chpl
- test/studies/hpcc/HPL/vass/schurComplement.chpl
- test/studies/hpcc/HPL/vass/hpl.hpcc2012.chpl

e-kayrakli commented:
Fixed the doc, modules and tests.

One question, though:

I didn't understand the motivation behind having a separate documentation for "local-on" pattern. There may be some optimizations and whatnot using an on statement tightly nested in a local block. But what is the purpose of documenting it separately? i.e. As a user, I am not sure what to learn different than the previous sections from the "local-on" section. Regardless, I changed "local-on" to "local-do-on" to be more in line with the new grammar.

benharsh commented:
Thanks for the changes, I'll fire off some testing.

I believe we have documented a 'local-do-on' separately because it will trigger an error if the destination of the on-statement is not on the same node. While this isn't useful for the flat locale model, it is more useful for numa.

e-kayrakli commented:
> I believe we have documented a 'local-do-on' separately because it will trigger an error if the destination of the on-statement is not on the same node. While this isn't useful for the flat locale model, it is more useful for numa.

Makes more sense now. Thanks

benharsh commented:
Testing came back clean! Ready for me to merge this?

e-kayrakli commented:
Good to go from my view.

benharsh commented:
I think it would make more sense for you to send the email as the main advocate for this change. I'm fine with holding this PR for a couple of days after the email is sent.

e-kayrakli commented:
@benharsh -- The mail is sent. You can merge this whenever you want.

benharsh commented:
@bradcray : it's been a few days with no responses on the mailing list, do you feel OK with merging this now?

bradcray commented:
I do, yes.
